olive/systems/python_environment/python_environment_system.py

```diff
@@ -0,0 +1,261 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+import os
+import pickle
+import tempfile
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import numpy as np
+import torch
+
+from olive.common.utils import run_subprocess
+from olive.evaluator.evaluation import compute_accuracy, compute_latency, format_onnx_input, get_user_config
+from olive.evaluator.metric import Metric, MetricType
+from olive.model import OliveModel, ONNXModel
+from olive.passes.olive_pass import Pass
+from olive.systems.common import Device, SystemType
+from olive.systems.olive_system import OliveSystem
+from olive.systems.system_config import PythonEnvironmentTargetUserConfig
+
+logger = logging.getLogger(__name__)
+
+
+class PythonEnvironmentSystem(OliveSystem):
+    system_type = SystemType.PythonEnvironment
+
+    def __init__(
+        self,
+        python_environment_path: Union[Path, str],
+        device: Device = Device.CPU,
+        environment_variables: Dict[str, str] = None,
+        prepend_to_path: List[str] = None,
+    ):
+        super().__init__()
+        self.config = PythonEnvironmentTargetUserConfig(
+            python_environment_path=python_environment_path,
+            device=device,
+            environment_variables=environment_variables,
+            prepend_to_path=prepend_to_path,
+        )
+        self.environ = deepcopy(os.environ)
+        if self.config.environment_variables:
+            self.environ.update(self.config.environment_variables)
+        if self.config.prepend_to_path:
+            self.environ["PATH"] = os.pathsep.join(self.config.prepend_to_path) + os.pathsep + self.environ["PATH"]
+        self.environ["PATH"] = str(self.config.python_environment_path) + os.pathsep + self.environ["PATH"]
+
+        # get available EPs
+        self.available_eps = self.get_available_eps()
+
```

Comment:
could we run the get available eps only when user not provide any EPs to save time?

Reply:
Done. Lazy loading of get_available_eps. It caches the values when it is run for the first time.